lib/matobjplist.gi

```diff
@@ -18,108 +18,90 @@
-    filter2 := filter and IsMutable;
+InstallMethod( NewVector, "for IsPlistVectorRep, a semiring, and a list",
+  [ IsPlistVectorRep, IsSemiring, IsList ],
+  function( filter, basedomain, l )
+    local typ;
+    filter := IsPlistVectorRep and IsMutable;
     if HasCanEasilyCompareElements(Representative(basedomain)) and
        CanEasilyCompareElements(Representative(basedomain)) then
-        filter2 := filter2 and CanEasilyCompareElements;
+        filter := filter and CanEasilyCompareElements;
     fi;
-    if IsIdenticalObj(basedomain,Integers) then
-        T := NewType(FamilyObj(basedomain),
-                       filter2 and IsIntVector);
+    if IsIdenticalObj(basedomain, Integers) then
+        filter := filter and IsIntVector;
     elif IsFinite(basedomain) and IsField(basedomain) then
-        T := NewType(FamilyObj(basedomain),
-                       filter2 and IsFFEVector);
-    else
-        T := NewType(FamilyObj(basedomain),
-                       filter2);
+        filter := filter and IsFFEVector;
     fi;
-    return T;
-  end);
-
-InstallMethod( NewVector, "for IsPlistVectorRep, a ring, and a list",
-  [ IsPlistVectorRep, IsRing, IsList ],
-  function( filter, basedomain, l )
-    local typ, v;
-    typ := MakePlistVectorType(basedomain,IsPlistVectorRep);
-    v := [basedomain,ShallowCopy(l)];
-    Objectify(typ,v);
-    return v;
+    typ := NewType(FamilyObj(basedomain), filter);
+    return Objectify(typ, [ basedomain, ShallowCopy(l) ]);
   end );
 
-InstallMethod( NewZeroVector, "for IsPlistVectorRep, a ring, and an int",
-  [ IsPlistVectorRep, IsRing, IsInt ],
+InstallMethod( NewZeroVector, "for IsPlistVectorRep, a semiring, and an int",
+  [ IsPlistVectorRep, IsSemiring, IsInt ],
   function( filter, basedomain, l )
-    local typ, v;
-    typ := MakePlistVectorType(basedomain,IsPlistVectorRep);
-    v := [basedomain,Zero(basedomain)*[1..l]];
-    Objectify(typ,v);
-    return v;
+    l := ListWithIdenticalEntries(l,Zero(basedomain));
+    return NewVector( filter, basedomain, l );
   end );
 
 InstallMethod( NewMatrix,
-  "for IsPlistMatrixRep, a ring, an int, and a list",
-  [ IsPlistMatrixRep, IsRing, IsInt, IsList ],
-  function( filter, basedomain, rl, l )
-    local nd, filterVectors, m, e, filter2, i;
-
+  "for IsPlistMatrixRep, a semiring, an int, and a list",
+  [ IsPlistMatrixRep, IsSemiring, IsInt, IsList ],
+  function( filter, basedomain, ncols, l )
+    local m;
     # If applicable then replace a flat list 'l' by a nested list
-    # of lists of length 'rl'.
-    if Length(l) > 0 and not IsVectorObj(l[1]) then
-      nd := NestingDepthA(l);
-      if nd < 2 or nd mod 2 = 1 then
-        if Length(l) mod rl <> 0 then
-          Error( "NewMatrix: Length of l is not a multiple of rl" );
+    # of lists of length 'ncols'.
+    if Length(l) = 0 then
+      # empty matrix
+      m := [];
+    elif IsVectorObj(l[1]) then
+      # list of vectors
+      # TODO: convert each IsVectorObj to a plist
```

Comment:
That sounds to me like you want to turn each entry of `l` into `IsPlistRep`. That can't be right, or am I misunderstanding something?

Reply:
A primary point of this PR is to change that, i.e.: do not store a list of vector objects, but rather store an old style matrix. This avoids a ton of overhead in both memory usage and performance. Before this PR, `IsPlistMatrixRep` was hopelessly slow compared to old-style matrices.